### PR TITLE
Support multiple providers with dynamic config

### PR DESCRIPTION
**Pull request by lawrencegripper**
This PR aims to allow multiple different providers and allow them to be configured from a configmap in k8s.

#16

### CODE REVIEW

**Comment by lawrencegripper** [comment]
image

Happy that this is good to review now.

File: README.md [hunk 1]
```diff
@@ -35,6 +35,45 @@ Users should be able to deploy the bridge into their cluster with a provider sel
 `ENABLE_PROVIDER_LOG` To enable full provider logs
 `SKIP_CRD_CREATION` To skip CRD creation at startup
 
+## Provider Config
+
+There are two ways to configure the provider used by the operator. 
+
+1. Normal provider environment variables
+2. An environment variable which contains the HCL for provider config
```
Comment (stuartleeks) [inline README.md:43]: It could be nice to also support a ConfigMap - what do you think? Maybe raise an issue for it?

File: main.go [hunk 2]
```diff
@@ -1,25 +1,23 @@
...
-	"github.com/hashicorp/terraform/plugin"
-	"github.com/hashicorp/terraform/plugin/discovery"
 	"k8s.io/client-go/rest"
 	"k8s.io/client-go/tools/clientcmd"
 	ctrl "sigs.k8s.io/controller-runtime"
 )
 
 func main() {
-	provider := getInstanceOfAzureRMProvider()
+	provider := getInstanceOfProvider("azurerm")
```
Comment (stuartleeks) [inline main.go:13]: Would it make sense for the provider name come from an environment variable or something like that?

File: provider.go [hunk 3]
```diff
@@ -23,37 +60,23 @@ func configureProvider(log logr.Logger, provider *plugin.GRPCProvider) {
...
-	featuresBlockMap := map[string]cty.Value{}
-	log.Info("%v", featuresType)
-	// Get each of the nested blocks required in the block and create
-	// empty items for them. Insert them into the featuresBlockMap
-	for name, nestedBlock := range featuresType.BlockTypes {
-		featuresBlockMap[name] = nestedBlock.EmptyValue()
+	providerConfigBlock := provider.GetSchema().Provider.Block
+
+	// Parse the content of the provider block given to us into a body.
+	file, diagParse := hclsyntax.ParseConfig([]byte(configBody), "test.tf", hcl.Pos{})
```
Comment (stuartleeks) [inline provider.go:66]: Is the filename (`test.tf`) important here? If not, how about `dummy.tf`?

File: provider.go [hunk 4]
```diff
@@ -62,24 +85,59 @@ func configureProvider(log logr.Logger, provider *plugin.GRPCProvider) {
 		Config: configFull,
 	})
 	if err := prepConfigResp.Diagnostics.Err(); err != nil {
-		log.Error(err, fmt.Sprintf("Failed to prepare config: %s", err))
-		panic("Failed to prepare config")
+		return nil, fmt.Errorf(`Failed to set configure provider from config: %w.`+
+			`Hint: See startup docs on using PROVIDER_CONFIG_HC or the providers envs to set required fields`, err)
```
Comment (stuartleeks) [inline provider.go:89]: nit
```suggestion
			`Hint: See startup docs on using PROVIDER_CONFIG_HC or the providers env vars to set required fields`, err)
```

File: provider.go [hunk 5]
```diff
@@ -62,24 +85,59 @@ func configureProvider(log logr.Logger, provider *plugin.GRPCProvider) {
...
-		panic(fmt.Sprintf("Failed to configure provider: %s", err))
+		return err
+	}
+
+	return nil
+}
+
+// This compliments the `emtypBlock` as it will check that blocks are correctly populated
+// when a single block is mandated (min 1 max 1)
+func populateSingleInstanceBlocks(value cty.Value, blocks map[string]*configschema.NestedBlock) cty.Value {
```
Comment (stuartleeks) [inline provider.go:115]: nice :-)

**Comment by lawrencegripper** [comment]
> It could be nice to also support a ConfigMap - what do you think? Maybe raise an issue for it?

Yeah, happy to look at how to improve this, I should have written up some more context on why I ended up with this approach. I'll also track as an issue.

The complexity with using a configMap comes due to the mismatch. On one side you have HCL with complex nested type system on the other you have configMap which is represented as a set of `KeyValue` pairs. When I started attempting to map between the two I ended up a) struggling b) making something very complex. 

The other option I looked at was a custom CRD but as discussed this creates a chicken and egg problem. The operator needs to run to create the CRD definition from the TF provider config but the operator wants to have the TF provider configured to be able to work. It would be possible to untangle this so that the operator installed the CRDs required then waited until the provider config CRD was set but the complexity of both the UX and the code felt wrong. 

The middle ground of msing the environment variables means they can be mounted from secrets or configmaps and the `hcl` env means additional, more complex, configuration can be provided for nested properties. 

Something like this would allow you to configure the `AzureRM` provider to work with the operator, for example. Not set on this as the final answer but it felt like a "good enough for now" approach. 

```
apiVersion: v1
kind: ConfigMap
metadata:
  name: arm-tfprovider-config
  namespace: default
data:
  ARM_CLIENT_SECRET: stuff
  ARM_CLIENT_SECRET: things
  PROVIDER_CONFIG_HCL: "features {}"
```

```
apiVersion: v1
kind: Pod
metadata:
  name: dapi-test-pod
spec:
  containers:
    - name: test-container
      image: k8s.gcr.io/busybox
      command: [ "/bin/sh", "-c", "env" ]
      envFrom:
      - configMapRef:
          name: arm-tfprovider-config
```

File: main.go (hunk 2 above)
Comment (lawrencegripper) [reply main.go:13]: Yes, I'm going to pick this up as #15

File: README.md (hunk 1 above)
Comment (lawrencegripper) [reply README.md:43]: I gave this one a go but turned out horribly, added more context in the PR thread below

File: provider.go (hunk 3 above)
Comment (lawrencegripper) [reply provider.go:66]: Great spot, yeah exactly right - will update to `dummy.tf`

**Comment by stuartleeks** [comment]
Thanks for the extra context :-)

I was thinking along the lines of [creating a `ConfigMap` from an HCL file](https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/#create-configmaps-from-files). E.g. have a file with the HCL for a `provider` block and embed that in the `ConfigMap`

But, as I said, I think this is food for future thought and this PR would be good to merge in :-)

**Comment by lawrencegripper** [comment]
Oh, wrong end of the stick entirely from me - Sorry! Yeah I like this idea a lot.

**Pull request merged**